### ISSUE TITLE
Bluesky Scans Freeze because it misses a signal from the network

### ISSUE THREAD

**Issue by stuartcampbell** [bluesky/bluesky#916]
This is a copy of the internal trac ticket (https://controlsweb.nsls2.bnl.gov/trac/ticket/1644) reported by @ambarb and captured by @johnsinsheimer

## Current Behavior
This problem happens now and again, but it has never affected this particular camera on subnet 1. The scan just hangs at a random position during the scan. It is difficult to align the beamline if this does not work reliably. Bluesky is waiting on a state to change, but it already changed and bluesky missed it.

This has been a nagging problem that we would like a solution for. It slows us down in aligning the beamline because we have to click motors by hand and imagine what the scan would like like on a graph.

Can someone please investigate to see what might be going on?

I also tried just a time scan on the camera itself (no moving hardware involved). This scan also hangs so I am inferring that it is related to the communication between the camera, epics, and the workstation.

**Comment by danielballan** [comment]
@johnsinsheimer What specifically is missed? The transition of the `Acquire` PV from high (acquiring) to low (done)?

**Comment by danielballan** [comment]
Possible solution: put the acquire PV on a high-priority VirtualCircuit. We've never tried this before. I'm not sure if/how pyepics exposes Channel Access's priority semantics.

**Comment by tacaswell** [comment]
It is also possible there is still a race-condition in how we handle AD.

**Comment by ambarb** [comment]
Later on, I figured out if I clicked, Start or Stop on the camera’s CSS page under Acquition, that the scan would move forward (it was in the start acquiring state before clicking).  In both cases, the plot in bluesky looked reasonable.  At some point, the camera did not flip back to Continuous mode when it exited the scan, but this was when I clicked “stop” in CSS. When are began clicking “Start”, the camera always went back to the pre-scan state  (it uses Multiple for the bsui scan).



Not sure if this helps, but it is an additional piece of information.

**Comment by danielballan** [comment]
I see. It seems to me that either:

* The camera got stuck in 'acquire' mode, and CSStudio and ophyd/bluesky are correctly reflecting that fact; or
* The IOC's message that it finished never reaches ophyd or CSStudio potentially due to network traffic.

It's not clear from what we know so far which is the case.

**Comment by ambarb** [comment]
When I use `msg_hook = print`, bsui seems to hang with “wait” as the last line.  If we can reproduce it ( or more likely, see it again), do you want use to past the `msg_hook` info here?  Is there any other information that would be useful for a root cause?  We've had the network looked at in the past, and no one seems to find anything obviously wrong.

Attn.: @cmazzoli @wen-hu @stuwilkins   DAMA may need more information from us.

**Comment by tacaswell** [comment]
The next time it happens can you caget the acquire PV?  If that agrees with CSS/ophyd that the bit is still high then it is a problem in the IOC, if it is low then it is a network issues with the monitor somehow getting lost to both process.

Might be worth also starting a camonitor on the aquire bit to see if anything weird is going on with it when it hangs.

**Comment by johnsinsheimer** [comment]
Andi just showed me where the scan hangs.  CS-Studio shows the detector is in the acquire state when it is hanging.  There were 2 errors in the IOC log recently, but the hanging happened after these errors, I don't think they are related.  But in case anyone wants to know the IOC error:

2017/11/14 16:24:03.971 prosilica:writeFloat64: error, status=18 function=71, name=ACQ_PERIOD, value=0.005000
2017/11/14 16:24:03.971 XF:23ID1-BI{Diag:5-Cam:1}cam1:AcquirePeriod devAsynFloat64 pPvt->result.status=18, process error

Again, this was before the hanging happened in the Bluesky scan.

Are you able to tell from scan error logging what the last command sent to the IOC was before it is hanging here?  The next step I'd like to try to figure out is if the IOC receives the message from the Bluesky scan, or what the states are of the various PVs to know where it goes wrong.

**Comment by ambarb** [comment]
Here are the olog recordings of what happened.

Screen shot of frozen scan.  The middle upper pane with `23-ID-1 End Station Diagnostic 2 – Cube` as the title is the camera in question.  You can see that the detector is in the acquire state.
https://logbook.nsls2.bnl.gov/23-ID-1/#64886_1 

Camera and nanopositioner scan  (relative_scan of nanopositioner)
https://logbook.nsls2.bnl.gov/23-ID-1/#64888_1


just count([cube_beam],num=None)
https://logbook.nsls2.bnl.gov/23-ID-1/#64889_1

**Comment by danielballan** [comment]
> Are you able to tell from scan error logging what the last command sent to the IOC was before it is hanging here?

No, the logbook doesn't provide enough granularity to be useful here. It's derived from Documents.

> The next step I'd like to try to figure out is if the IOC receives the message from the Bluesky scan, or what the states are of the various PVs to know where it goes wrong.

This is where plugging in caproto would be useful, where we could literally see at a byte level the last thing we told the IOC and the last thing it said back. (It's impossible to get that level of detailed logging from pyepics; you'd have to modify libca itself.) Unfortunately ophyd + caproto doesn't quite work yet, but this is as good a reason as any to get it working. Ken and I worked on that today and I intend to put about 50% of my time into it until it works.

**Comment by brunoseivam** [comment]
This might be of interest:

http://www.aps.anl.gov/epics/tech-talk/2016/msg00714.php

In my experience when in Multiple mode the Prosilica driver can miss a frame and be stuck in the Acquire state, holding up the acquisition. Following the advice on the aforementioned thread made the problem go away (at least at FMX/AMX):

    sudo sysctl -w net.core.rmem_max=1048576 net.core.rmem_default=1048576

**Comment by danielballan** [comment]
Thanks, @brunoseivam!

**Comment by hhslepicka** [comment]
Just adding to Bruno's answer, it is also good to increase the MTU on the ethernet card connected to the camera:
```bash
if [[ "`uname -r`" == *".el7."* ]]; then
	ifconfig eno1 mtu 9000
else
	ifconfig eth0 mtu 9000
fi
```

**Comment by cowanml** [comment]
mtu is 9000 on all cam iocs, but looks like rmem settings could be improved...
-matt


cowan@box64-1:~$ pdsh -l root -w xf23id1-ioc[1,3],xf23id2-ioc1,xf23ida-ioc1 ifconfig|grep 'HWaddr\|inet addr\|MTU'|grep -C1 10\.23\.1|less -XMi
xf23id1-ioc3: eth5      Link encap:Ethernet  HWaddr 40:f2:e9:18:73:34
xf23id1-ioc3:           inet addr:10.23.1.5  Bcast:10.23.1.255  Mask:255.255.255.0
xf23id1-ioc3:           UP BROADCAST RUNNING MULTICAST  MTU:9000  Metric:1
--
xf23ida-ioc1: eth2      Link encap:Ethernet  HWaddr 40:f2:e9:cd:5d:b4
xf23ida-ioc1:           inet addr:10.23.1.1  Bcast:10.23.1.255  Mask:255.255.255.0
xf23ida-ioc1:           UP BROADCAST RUNNING MULTICAST  MTU:9000  Metric:1
--
xf23id1-ioc1: eth1      Link encap:Ethernet  HWaddr 90:e2:ba:50:d9:3d
xf23id1-ioc1:           inet addr:10.23.1.2  Bcast:10.23.1.255  Mask:255.255.255.0
xf23id1-ioc1:           UP BROADCAST RUNNING MULTICAST  MTU:9000  Metric:1
--
xf23id2-ioc1: eth7      Link encap:Ethernet  HWaddr 90:e2:ba:53:b3:b4
xf23id2-ioc1:           inet addr:10.23.1.3  Bcast:10.23.1.255  Mask:255.255.255.0
xf23id2-ioc1:           UP BROADCAST RUNNING MULTICAST  MTU:9000  Metric:1

cowan@box64-1:~$ pdsh -l root -w xf23id1-ioc[1,3],xf23id2-ioc1,xf23ida-ioc1 'sysctl -a|grep net\.core\.rmem'|dshbak -c
----------------
xf23id1-ioc1,xf23id2-ioc1,xf23ida-ioc1
----------------
net.core.rmem_max = 131071
net.core.rmem_default = 229376
----------------
xf23id1-ioc3
----------------
net.core.rmem_max = 2147483647
net.core.rmem_default = 229376

**Comment by cowanml** [comment]
fyi, on any nodes participating in gpfs we typically set
     net.core.rmem_max = 16777216  (eg. 16MB [16*1024^2])
but doesn't look like this is the case on xf23id1-ioc1 or xf23id1-ioc3 ?

The originally linked ptgrey kb article
(http://www.ptgrey.com/KB/10016) also suggest:
     "Increase packet delay time using the FlyCapture2 API or the FlyCap2 program."
although they suggest what to increase it to?

Not sure how xf23id1-ioc3 ended up with rmem_max==2GB ?

I've also seen suggestions net.ipv4.tcp_mtu_probing should be '1' when
jumbo frames (mtu > 1500) are enabled, seems reasonable.

hrm... there's actually conflicting suggestions about tcp_sack, ibm
(gpfs) says '0', but mellanox (infiniband and 40gb ethernet) says '1'.

-matt


For comparison... nsd-001 shows typical settings for gpfs nodes
----------------------------------------------------------------------
cowan@box64-1:~$ pdsh -l root -w nsd-001,xf23id1-ioc[1,3],xf23id2-ioc1,xf23ida-ioc1 'sysctl -a|grep '\''[\._][rw]mem\|netdev\.max_backlog\|mtu_prob\|tcp_time\|tcp_sack\|low_late\|optmem_max\|tcp_mem'\''|sort'|dshbak -c|less -XMi
----------------
xf23id1-ioc1,xf23id2-ioc1
----------------
net.core.optmem_max = 20480
net.core.rmem_default = 229376
net.core.rmem_max = 131071
net.core.wmem_default = 229376
net.core.wmem_max = 131071
net.ipv4.tcp_low_latency = 0
net.ipv4.tcp_mem = 383475       511301  766950
net.ipv4.tcp_mtu_probing = 0
net.ipv4.tcp_rmem = 4096        87380   6291456
net.ipv4.tcp_sack = 1
net.ipv4.tcp_timestamps = 1
net.ipv4.tcp_wmem = 4096        16384   4194304
----------------
xf23ida-ioc1
----------------
net.core.optmem_max = 20480
net.core.rmem_default = 229376
net.core.rmem_max = 131071
net.core.wmem_default = 229376
net.core.wmem_max = 131071
net.ipv4.tcp_low_latency = 0
net.ipv4.tcp_mem = 383451       511269  766902
                                    ^^^     ^^^  only diffs compared to above hosts
net.ipv4.tcp_mtu_probing = 0
net.ipv4.tcp_rmem = 4096        87380   6291456
net.ipv4.tcp_sack = 1
net.ipv4.tcp_timestamps = 1
net.ipv4.tcp_wmem = 4096        16384   4194304
----------------
nsd-001
----------------
net.core.optmem_max = 16777216    << diffs compared to -ioc's above
net.core.rmem_default = 16777216    <<
net.core.rmem_max = 16777216    <<
net.core.wmem_default = 16777216    <<
net.core.wmem_max = 16777216    <<
net.ipv4.tcp_low_latency = 1    <<
net.ipv4.tcp_mem = 16777216     16777216        16777216    <<
net.ipv4.tcp_mtu_probing = 0
net.ipv4.tcp_rmem = 4096        87380   16777216    <<
net.ipv4.tcp_sack = 0    <<
net.ipv4.tcp_timestamps = 0    <<
net.ipv4.tcp_wmem = 4096        87380   16777216    <<
----------------
xf23id1-ioc3
----------------
net.core.optmem_max = 20480
net.core.rmem_default = 229376
net.core.rmem_max = 2147483647    << diffs compared to -ioc's above
net.core.wmem_default = 229376
net.core.wmem_max = 2147483647    <<
net.ipv4.tcp_low_latency = 0
net.ipv4.tcp_mem = 6204150      8272203 12408300    <<
net.ipv4.tcp_mtu_probing = 1    <<
net.ipv4.tcp_rmem = 4096        87380   6291456
net.ipv4.tcp_sack = 1
net.ipv4.tcp_timestamps = 1
net.ipv4.tcp_wmem = 4096        16384   4194304
----------------------------------------------------------------------

**Comment by cowanml** [comment]
I suggest we try:
----------------------------------------------------------------------
# default 1MB, max 4MB
# stock=229376,131071  (yup, default>max wtf?)
sudo sysctl -w net.core.rmem_default=1048576
sudo sysctl -w net.core.rmem_max=4194304

# do the same for transmitting packets... ioc also sends video to wkstns...
sudo sysctl -w net.core.wmem_default=1048576
sudo sysctl -w net.core.wmem_max=4194304

# disable prequeue
sudo sysctl -w net.ipv4.tcp_low_latency=1
# enable mtu detection when icmp blackhole detected
sudo sysctl -w net.ipv4.tcp_mtu_probing=1

# min/default/max tcp buffer sizes
# Make wmem default match rmem default.
# Default overrides net.core.[rw]mem_default
# Max *DOES NOT* override net.core.[rw]mem_max
# Wow, there are alot of inconsistencies in the stock settings!  Like
# rmem_max<rmem_default above or the max below can't be hit unless you
# bumped net.core.rmem_max to this or higher?
sudo sysctl -w net.ipv4.tcp_wmem="4096        87380   6291456"
----------------------------------------------------------------------

That can be done anytime by anyone with sudo.  The change should
immediately take effect for new connections (will probably need to
restart the ioc), and should be pretty innocuous, except for hopefully
solving this problem.

To revert, just run the same command with the previous value (recorded
earlier in email thread below).  If they work and don't cause other
problems, we can make them persistent across a reboot by adding them
to /etc/sysctl.conf and consider propagating them to all camera iocs
(with care on xf23id1-ioc3, which seems to already have gotten some
tweaks).

If it doesn't solve the problem, or causes other problems, I suggest
reverting tcp_low_latency=0 and tcp_mtu_probing=0 and retrying before
completely reverting.

-matt

**Comment by danielballan** [comment]
Would someone be willing to document these suggested settings somewhere in ophyd's documentation?

Also, as an internal NSLS-II concern, could we create a puppet class or ansible playbook encoding proper configuration for AD IOCs?

**Comment by cowanml** [comment]
On Thu, 16 Nov 2017, Dan Allan wrote:
> 
> Would someone be willing to document these suggested settings somewhere in ophyd's
> documentation?
> 
> Also, as an internal NSLS-II concern, could we create a puppet class or ansible playbook
> encoding proper configuration for AD IOCs?


I'm not sure it's optimal to generalize this to all AD IOCs A safer
bet would be standard gigE camera iocs with jumbo frames interfaces on
the camera vlan.

The settings *might* benefit other AD ioc types, but that needs to be evaluated 
for each detector.

This working optimally also requires the cameras (and other detectors)
to be configured to match (eg. jumbo frames)


yup, puppet or ansible is the only way to ensure this is done consistently.
-matt

**Comment by stuartcampbell** [comment]
Can I suggest that during the next maintenance day everyone sits down together at CSX and makes sure that these network changes actually fixes the issue ?

**Comment by cowanml** [comment]
If it's regularly problematic I'll pop by the beamline tomorrow and
see if Claudio wants to try it.  It's quick/minimally disruptive and
easily revertable, so might be worth trying even with users.

-matt

**Comment by stuwilkins** [comment]
@cowanml, thanks for the offer. However, I would prefer doing this all together if possible in a controlled way so that we can find the route cause. It would be great if you could be there with the DAMA group to troubleshoot this in downtime.


On Nov 16, 2017, at 5:48 PM, cowanml <notifications@github.com<mailto:notifications@github.com>> wrote:


If it's regularly problematic I'll pop by the beamline tomorrow and
see if Claudio wants to try it. It's quick/minimally disruptive and
easily revertable, so might be worth trying even with users.

-matt

**Comment by cowanml** [comment]
On Fri, 17 Nov 2017, Stuart Wilkins wrote:
> @cowanml, thanks for the offer. However, I would prefer doing this
> all together if possible in a controlled way so that we can find the
> route cause. It would be great if you could be there with the DAMA
> group to troubleshoot this in downtime.

ok, will be happy to be there, just let me know when.

If it becomes more problematic anyone with sudo can easily try/revert
it anytime, also need to restart the ioc after making the changes with
sysctl.

-matt

**Comment by danielballan** [comment]
Did this meeting ever occur?

**Comment by ambarb** [comment]
No.  I am not sure if anyone is actually taking point on this.
@stuwilkins?

**Comment by ambarb** [comment]
@cmazzoli agreed to organize.  @danielballan did you want to participate?

**Comment by danielballan** [comment]
I'm running out of in-person time this week but might be able to assist remotely

**Comment by ambarb** [comment]
@cowanml and @johnsinsheimer  we are also having these troubles at 2ID.  Can we test the possible solution here?   Will tomorrow work or do we need to wait until next week?

**Comment by cowanml** [comment]
As mentioned previously, the solution should be simple quick, and easily revertable by anyone with sudo, so I'm confident recommending trying it, but it's a beamline operations call (see previous messages from others).  In any case, I'm available...
-matt

**Comment by stuwilkins** [comment]
@ambarb @johnsinsheimer these settings were changed on IOC2 at 2-ID. I don't know if the camera IOC was rebooted to make sure it opened with a bigger buffer. 

However, A long time ago at CSX we worked out that you have to put the camera IOC on 10G for the 10.2.1.0/24 subnet becuase if you have more than one camera it is very easy to saturate that link. 

Was this ever taken care of? @cowanml 2-ID is also waiting for the whole operator area to be "upped" to 10G and also for a 10G upgrade to the core.

**Comment by ambarb** [comment]
we are now have trouble with the  new cameras added for the new measurements.  @johnsinsheimer upgraded these new docs to AD version 3 so these things were definitely restarted a few times yesterday.   

I increased the acquire period such that period = exposure_time +0.5.  And then I made all but one camera have a 15s acquire period.  This reduced the occurrences but it still happens so you cannot walk away from a scan.

This might point to stuart's point of starting the link.

**Comment by ambarb** [comment]
@cowanml can you let us know what kind of bandwidth we have for ioc2 ate the 2id beamline?

**Comment by cowanml** [comment]
All interfaces on the cam ioc server are just 1g.

From linux side no dropped packets or anything bad like that, but
looking at the switch side, Leon does see a sizeable number of dropped
packets (observably incrementing) on the port for xf02id1-ioc2's cam
network interface.  Network utilization on the port was just 5%
though?!  We looked at cacti and it looks like the 2nd member of the
net-02id1-agg stack isn't showing up in cacti, so we were only able to
look at the instantaneous utilization, not history.

Would make sense to me to bump the cam network interface up to 10g (it
is pulling from 10x 1g cams...), but that will have to wait for the
new 10g switch Leon is working on and installation of a 10g card and ...

No reason not to proceed with the vendor recommended network stack
tuning in the meantime.
-matt

----------------------------------------------------------------------
cowan@ssh01:~$ pdsh -l root -g 'bl=2&&class=ioc' 'manage-iocs status|grep cam'
xf02id1-ioc2: /etc/init.d/softioc-cam01-AD3-1           Running
xf02id1-ioc2: /etc/init.d/softioc-cam02-AD3-1           Running
xf02id1-ioc2: /etc/init.d/softioc-cam03-AD3-1           Running
xf02id1-ioc2: /etc/init.d/softioc-cam04-AD3-1           Running
xf02id1-ioc2: /etc/init.d/softioc-cam05-AD3-1           Running
xf02id1-ioc2: /etc/init.d/softioc-cam07-AD3-1           Running
xf02id1-ioc2: /etc/init.d/softioc-cam08-AD3-1           Running
xf02id1-ioc2: /etc/init.d/softioc-cam09-AD3-1           Running
xf02id1-ioc2: /etc/init.d/softioc-cam10-AD3-1           Running
xf02id1-ioc2: /etc/init.d/softioc-cam15-AD3-1           Running

root@xf02id1-ioc2:~# route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         gw-210.cs.nsls2 0.0.0.0         UG    0      0        0 eth0
localnet        *               255.255.255.0   U     0      0        0 eth0
10.2.1.0        *               255.255.255.0   U     0      0        0 eth1
10.2.2.0        *               255.255.254.0   U     0      0        0 eth2

root@xf02id1-ioc2:~# route|sed -n 's|^.*[       ]\([a-z0-9]*[0-9]\)[    ]*$|\1|p'|sort -u|xargs -ri bash -c "echo -n {}; ethtool {} |grep -i speed"
eth0    Speed: 1000Mb/s
eth1    Speed: 1000Mb/s
eth2    Speed: 1000Mb/s
----------------------------------------------------------------------

**Comment by cowanml** [comment]
If any of the cam feeds aren't strictly necessary at a given point in
time, you could always stop those iocs to eliminate wasting the
currently limited bandwidth?

I wish area detector was smart enough to know when nothing was
consuming a feed or a PV generated from it and not pull video from a
camera when it didn't need to :(

-matt

**Comment by cowanml** [comment]
my previous inlined commands+output didn't make it clear the
interfaces in xf02id1-ioc2 are all only 1g capable, just showed they
were currently running at 1g, but they are all indeed only 1g capable.

-matt

----------------------------------------------------------------------
root@xf02id1-ioc2:~# ifconfig -a | sed -n 's|^\([^      ][a-z0-9]*\)[   ].*$|\1|p'|grep -v '^\(lo\|usb[0-9]\)$' | xargs -r -n1 ethtool |grep -A3 '^S\|Supported link modes'
Settings for eth0:
         Supported ports: [ TP ]
         Supported link modes:   10baseT/Half 10baseT/Full
                                 100baseT/Half 100baseT/Full
                                 1000baseT/Full
         Supported pause frame use: Symmetric
--
Settings for eth1:
         Supported ports: [ TP ]
         Supported link modes:   10baseT/Half 10baseT/Full
                                 100baseT/Half 100baseT/Full
                                 1000baseT/Full
         Supported pause frame use: Symmetric
--
Settings for eth2:
         Supported ports: [ TP ]
         Supported link modes:   10baseT/Half 10baseT/Full
                                 100baseT/Half 100baseT/Full
                                 1000baseT/Full
         Supported pause frame use: Symmetric
--
Settings for eth3:
         Supported ports: [ TP ]
         Supported link modes:   10baseT/Half 10baseT/Full
                                 100baseT/Half 100baseT/Full
                                 1000baseT/Full
         Supported pause frame use: Symmetric
----------------------------------------------------------------------

**Comment by cowanml** [comment]
On Thu, 22 Feb 2018, Matt Cowan wrote:
...
> Would make sense to me to bump the cam network interface up to 10g (it
> is pulling from 10x 1g cams...), but that will have to wait for the
> new 10g switch Leon is working on and installation of a 10g card and ...
...

ignore the blurb about "new 10g switch" that was me mixing up
unrelated work on a different beamline.

We would need a 10g nic+optics for xf02id1-ioc2 and I don't know what
the status of available 10g ports at the beamline is?

-matt

**Comment by stuwilkins** [comment]
@cowanml This was discussed with Rob many months ago. Could you please look at the stwitches and IOCs and advise what is needed. Thanks.

S

**Comment by cowanml** [comment]
On Thu, 22 Feb 2018, Stuart Wilkins wrote:
> 
> @cowanml This was discussed with Rob many months ago. Could you please look at the stwitches
> and IOCs and advise what is needed. Thanks.
> 
> S


I'll look into the switches with Leon.


How did we go from a recommendation of
     net.core.[rw]mem_default=1048576  # 1MB
     net.core.[rw]mem_max=4194304  # 4MB

to
     net.core.[rw]mem_{default,max} = 2147483647 ?  # 2GB!

xf23id1-ioc3 was like this when I looked at it back in Nov, not sure about
xf23id1-ioc3 or xf02id1-ioc2, but all 3 are set at 2GB now.


That's a huge jump, leaves no room between _default and _max for the
system to adjust itself, and likely to consume excessive amounts of
memory with marginal increased benefit, could very well even decrease
performance:  https://en.wikipedia.org/wiki/Bufferbloat

As a datapoint, the vendor recommended setting for gpfs is just 16MB.


Looks like xf23id2-ioc1 and xf23ida-ioc1 still have the stock
     net.core.[rw]mem_default = 229376
     net.core.[rw]mem_max = 131071

and could use
     net.core.[rw]mem_default = 1048576
     net.core.[rw]mem_max = 4194304

-matt

**Comment by danielballan** [comment]
Would it be alright if I closed this issue as out of scope for bluesky, as it seems to be a facility-specific IT concern?

**Comment by stuwilkins** [comment]
If we could transfer it to a similar system I would be happy, but would you mind keeping it here so we can be a record in one place of what fixes this? Thanks.

**Comment by stuwilkins** [comment]
I spent a lot of time tuning this for our fast camera. There is a difference also for GPFS and this because GPFS does not rely on UDP but rather has a handshaking protocol *and* does not run in user space as far as I know. This is in fact the root of the problem. The camera streams in UDP without handshaking which is fairly common. If the packets get dropped or the buffer overruns in the kernel the image is lost.....

xf02ida-ioc1 and xf02id2-ioc1 were not changed as I was told there were no cameras on them.

For this application the max has to equal the default because there is no way to adjust the socket buffer at the application layer. I am looking into that now. I could do it when I wrote the low level driver for the fast CCD, but I will look to see if we can do this for the Prosilica. In the mean time that is why max = default.

We have plenty of spare 10G cards if these ioc’s were not purchased with 10 G cards in the first place.

@cowanml, should we move this to TRAC?

On Feb 23, 2018, at 9:39 AM, cowanml <notifications@github.com<mailto:notifications@github.com>> wrote:

On Thu, 22 Feb 2018, Stuart Wilkins wrote:
>
> @cowanml This was discussed with Rob many months ago. Could you please look at the stwitches
> and IOCs and advise what is needed. Thanks.
>
> S


I'll look into the switches with Leon.


How did we go from a recommendation of
net.core.[rw]mem_default=1048576 # 1MB
net.core.[rw]mem_max=4194304 # 4MB

to
net.core.[rw]mem_{default,max} = 2147483647 ? # 2GB!

xf23id1-ioc3 was like this when I looked at it back in Nov, not sure about
xf23id1-ioc3 or xf02id1-ioc2, but all 3 are set at 2GB now.


That's a huge jump, leaves no room between _default and _max for the
system to adjust itself, and likely to consume excessive amounts of
memory with marginal increased benefit, could very well even decrease
performance: https://en.wikipedia.org/wiki/Bufferbloat

As a datapoint, the vendor recommended setting for gpfs is just 16MB.


Looks like xf23id2-ioc1 and xf23ida-ioc1 still have the stock
net.core.[rw]mem_default = 229376
net.core.[rw]mem_max = 131071

and could use
net.core.[rw]mem_default = 1048576
net.core.[rw]mem_max = 4194304

-matt

**Comment by cowanml** [comment]
good info!  I forwarded this to a new Trac ticket
(https://controlsweb.nsls2.bnl.gov/trac/ticket/1888)

Does that mean Dan can close this bluesky issue?
Or you want to keep it open until the issue is proven to be resolved?
-matt

**Comment by danielballan** [comment]
I try to keep in mind that bluesky issues also spam our external collaborators, but I hear @stuwilkins request for keeping an easy-to-follow thread in this case, so let's resolve the issue here. Sorry for muddying the waters, @stuwilkins.

**Comment by stuwilkins** [comment]
Lets, close it, I will reopen and post the results at the end.

**Comment by prjemian** [comment]
I don't mind listening to the conversations.  At some point, it might be educational here.

**Comment by danielballan** [comment]
Fair enough! Thanks for the chiming in, @prjemian.

**Comment by danielballan** [comment]
Closing as stale and not a bluesky bug.
